src/rgw/rgw_admin.cc

```diff
@@ -5229,6 +5233,112 @@ int main(int argc, char **argv)
    }
   }
 
+  if (opt_cmd == OPT_BUCKET_RECALC_STATS) {
+    if (bucket_name.empty()) {
+      cerr << "ERROR: bucket not specified" << std::endl;
+      return -EINVAL;
+    }
+
+    if (!yes_i_really_mean_it) {
+      cerr << "This command will update bucket stats and object stats" << std::endl;
+      cerr << "do you really mean it? (requires --yes-i-really-mean-it)" << std::endl;
+      return 1;
+    }
+
+    rgw_bucket bucket;
+    RGWBucketInfo bucket_info;
+    map<string, bufferlist> attrs;
+    int ret = init_bucket(tenant, bucket_name, bucket_id, bucket_info, bucket, &attrs);
+    if (ret < 0) {
+      cerr << "ERROR: could not init bucket: " << cpp_strerror(-ret) << std::endl;
+      return -ret;
+    }
+
+    cout << "tenant: " << bucket_info.bucket.tenant << std::endl;
+    cout << "bucket name: " << bucket_info.bucket.name << std::endl;
+    cout << "bucket instance id: " << bucket_info.bucket.bucket_id << std::endl;
+
+    /* update bucket info  -- in progress*/
+    list<rgw_cls_bi_entry> entries;
+
+    if (max_entries < 0) {
+      max_entries = 1000;
+    }
+
+    if (verbose) {
+      formatter->open_array_section("entries");
+    }
+
+    uint64_t total_entries = 0;
+
+    int num_source_shards = (bucket_info.num_shards > 0 ? bucket_info.num_shards : 1);
+    string marker;
+    map<uint8_t, rgw_bucket_category_stats> stats;
+
+    for (int i = 0; i < num_source_shards; ++i) {
+      bool is_truncated = true;
+      marker.clear();
+      while (is_truncated) {
+	entries.clear();
+	ret = store->bi_list(bucket, i, string(), marker, max_entries, &entries, &is_truncated);
+	if (ret < 0 && ret != -ENOENT) {
+	  derr << "ERROR: bi_list(): " << cpp_strerror(-ret) << dendl;
+	  return -ret;
+	}
+
+	list<rgw_cls_bi_entry>::iterator iter;
+	for (iter = entries.begin(); iter != entries.end(); ++iter) {
+	  rgw_cls_bi_entry& entry = *iter;
+
+	  if (verbose) {
+	    formatter->open_object_section("entry");
+	    encode_json("shard_id", i, formatter);
+	    encode_json("num_entry", total_entries, formatter);
+	    encode_json("entry", entry, formatter);
+	  }
+	  total_entries++;
+
+	  if (entry.type == PlainIdx) {
+	    rgw_bucket_dir_entry dir_entry;
+	    auto bi_iter = entry.data.begin();
+	    ::decode(dir_entry, bi_iter);
+	    if (dir_entry.meta.accounted_size == 0 && dir_entry.meta.size != 0) {
```

Comment:
This was caused by a dir_suggest issue that would cause to complete op not to run. The simple way to produce is running s3cmd sync <folder> bucket> with more than 4000 files and in parallel run s3cmd ls -r <bucket> in a loop.